Added flow marking functionality in the console

drahosj opened the pull request:
Some functionality in https://github.com/mitmproxy/mitmproxy/issues/623 was actually fairly trivial to implement. The following features are included

In the flow list view:
- Press 'm' to toggle an easily visible mark on the current flow.
- When the flow list is cleared with 'C', marked flows are not deleted.
- If all flows in the list are marked, marked flows are deleted with 'C'

This means that pressing C twice will clear all flows, including marked ones. 

The mark is made large to stand out while scrolling, but that can be changed by changing SYMBOL_MARK in libmproxy/console/common.py.

cortesi commented:
Hi there. A very useful patch! I'd like one change before I merge this. I want to keep the protocol definitions as clean as possible, dealing only the actual protocols, and not with data used by the various front-ends. As it happens, mitmproxy console already has a mechanism for keeping "per flow" settings in a clean way - look at the flow settings functions [here](https://github.com/mitmproxy/mitmproxy/blob/master/libmproxy/console/__init__.py#L37-L43).

So, could you please:

- Remove the changes to libmproxy/protocol/primitives.py
- Move the tagging into a flow setting on the console app state instead

After that, I'll merge, and we can chat about how to take the functionality forward. I'm hesitant about having the C shortcut effectively mean two different things depending on a context that can change very rapidy in a live proxy session. I think we'd be better served with an "unmark all" shortcut or something similar. But these are details we can discuss once we have this in base. Thanks!

drahosj commented:
I made the changes you requested and moved all of the marking functionality out of the primitive and into the console. I also implemented "unmark all" on U instead of double-clear. It does make more sense to work that way.

Complete list of marking functionality:
- m in FlowList to toggle mark on the flow
- U in FlowList to unmark all flows
- C clears all unmarked flows, marked flows remain
- w now has an 'm' option to write all marked flows to a file
- When a marked flow is duplicated, the new copy will not be marked. This facilitates an easy duplicate, modify, replay workflow.

cortesi commented:
Excellent, thanks!

cortesi commented:
@drahosj: If you wanted to continue to polish the marking implementation, there are a few improvements that could be made. 

First, it would be useful to be able to mark/unmark from within a flow view. We should use the same keybinding, so "m" would do there too. At the moment, "m" is used to change the display mode in that view - we can move that to "M" instead. The help document erroneously claims that "M" is used to change the default display mode - that's now actually done in the options screen.

Second, we need a "marked" indicator in the flow view itself as well. 

Third, I don't love the fact that the marker is pre-pended in the flow list. It changes the alignment, and makes things jump around. I'd prefer that to be stable. I don't have a firm suggestion for how to do this - we could sacrifice a column to have the spacing there for all flows, we could integrate it into the current flow indicator column somehow. Something to think about.

Thanks again for the contribution - very handy indeed.
